draft-ietf-httpbis-safe-method-w-body.xml

```diff
@@ -238,42 +238,90 @@ q=foo&amp;limit=10&amp;sort=-published
+      that is, the client does not request or expect any change to the
+      state of the target resource. This does not prevent the server
+      from creating additional HTTP resources through which additional
+      information can be retrieved (see Sections <xref target="content-location" format="counter"/>
+      and <xref target="location" format="counter"/>).
     </t>
     <t>
-      A successful response to a QUERY request is expected to provide some
-      indication as to the final disposition of the operation. For
-      instance, a successful query that yields no results can be represented
-      by a 204 (No Content, <xref target="HTTP" section="15.3.5"/>) response. If the response includes content,
-      it is expected to describe the results of the operation.
+      Furthermore, QUERY requests are idempotent
+      (<xref target="HTTP" sectionFormat="comma" section="9.2.2"/>) --
+      they can be retried or repeated when needed, for instance after
+      a connection failure.
     </t>
+    <t>
+      As per <xref target="HTTP" section="15.3"/>, a 2xx (Successful) response code signals that
+      the request was successfully received, understood, and accepted.
+    </t>
+    <t>
+      In particular, a 200 (OK) response indicates that the query
+      was successfully processed and the results of that processing are
+      enclosed as the response content.
+    </t>
+
+    <section title="Media Types and Content Negotiation" anchor="media.types.and.conneg">
+      <t>
+        The semantics of a QUERY request depends both on the request content and the associated
+        metadata, such as the Media Type (<xref target="HTTP" sectionFormat="comma" section="8.3.1"/>).
+        In general, any problem with requests where content and metadata are inconsistent &MUST; be
+        rejected with a 4xx (Client Error) response (<xref target="HTTP" sectionFormat="comma" section="15.5"/>).
+      </t>
+      <t>
+        The list below describe various cases of failures and recommends specific status codes:
+      </t>
+      <ul>
+        <li>
+          A request lacking media type information by definition is incorrect and needs to fail
+          with a 4xx status code such as 400 (Client Error).
+        </li>
+        <li>
+          If a media type is specified, but not supported by the resource, a 415 (Unsupported Media Type)
+          is appropriate. This specifically includes the case where the media type is
+          known in principle, but lacks semantics specific to a QUERY to the target resource. In both cases,
+          the Accept-Query response field (<xref target="field.accept-query"/>) can be used for inform
+          the client of media types which are supported.
+        </li>
+        <li>
+          If a media type is specified, but is inconsistent with the actual request content,
+          a 400 (Bad Request) can be returned. Thus, "content sniffing" is not allowed.
```

Comment:
Fine with the error, but I'm unclear on the "content sniffing is not allowed." Isn't it content-sniffing to inspect the content and determine that it's not the claimed content type? Or are you meaning that a server isn't allowed to infer the correct content-type from the payload and override a missing/erroneous content-type header?

Assuming it's the latter, I'd change it to something like "The server MUST NOT attempt to infer the correct media type."

Reply:
The latter. Will do.